engine/src/main/java/org/terasology/rendering/dag/nodes/DownSamplerNode.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.dag.nodes;
+
+import org.terasology.assets.ResourceUrn;
+import org.terasology.monitoring.PerformanceMonitor;
+import org.terasology.rendering.assets.material.Material;
+import org.terasology.rendering.dag.ConditionDependentNode;
+import org.terasology.rendering.dag.stateChanges.BindFBO;
+import org.terasology.rendering.dag.stateChanges.EnableMaterial;
+import org.terasology.rendering.dag.stateChanges.SetInputTextureFromFBO;
+import org.terasology.rendering.dag.stateChanges.SetViewportToSizeOf;
+import org.terasology.rendering.opengl.BaseFBOsManager;
+import org.terasology.rendering.opengl.FBO;
+import org.terasology.rendering.opengl.FBOConfig;
+import org.terasology.rendering.opengl.FBOManagerSubscriber;
+
+import static org.terasology.rendering.dag.stateChanges.SetInputTextureFromFBO.FboTexturesTypes.ColorTexture;
+import static org.terasology.rendering.opengl.OpenGLUtils.renderFullscreenQuad;
+
+/**
+ * Instances of this class take the content of the color attachment of an input FBO
+ * and downsamples it into the color attachment of a smaller output FBO.
+ */
+public class DownSamplerNode extends ConditionDependentNode implements FBOManagerSubscriber {
+
+    private static final String TEXTURE_NAME = "tex";
+    private static final int SLOT_0 = 0;
+    private static final ResourceUrn DOWN_SAMPLER_MATERIAL = new ResourceUrn("engine:prog.downSampler");
+
+    private String performanceMonitorLabel;
+    private BaseFBOsManager outputFBOmanager;
+    private FBOConfig outputFBOconfig;
+    private FBO outputFBO;
+    private Material downSampler;
+
+    /**
+     * Throws a RuntimeException if invoked. Use initialise(...) instead.
+     */
+    @Override
+    public void initialise() {
+        throw new RuntimeException("Please do not use initialise(). For this class use initialise(...) instead.");
```

Comment:
Yes, it's not nice. There was some thought behind it, in the sense that this was the best either I could come up with or both tdgunes and I could come up with - I don't remember at what stage it was introduced.

I think it has to do with the fact that the superclass has an initialise() method that needs to be overridden, but this particular class (and others like it) need initialization parameters that vary from class to class. In this context, to create a NodeFactory specific for each of these type of nodes seemed overkill and we settled for a slightly improved NodeFactory capable of delaying initialization if necessary. This allow for a class-specific initialise(...) method to be called after the NodeFactory has instantiated (but not initialised) a node. The WorldRendererImpl class has a number of example usages of this in the initRenderGraph method.

Ultimately I'd welcome suggestion: it's not exactly elegant.